library/init/meta/tactic.lean

```diff
@@ -270,6 +270,34 @@ meta def decorate_ex (msg : format) (t : tactic α) : tactic α :=
 @[inline] meta def read : tactic tactic_state :=
 λ s, success s s
 
+/--
+`capture t` acts like `t`, but succeeds with a result containing either the returned value
+or the exception.
+
+The result can be used to inspect the error message, or passed to `unwrap` to rethrow the
+failure later.
+-/
+meta def capture (t : tactic α) : tactic (tactic_result α) :=
+λ s, match t s with
+| (success r s') := success (success r s') s'
+| (exception f p s') := success (exception f p s') s'
+end
+
+/--
+`unwrap r` unwraps a result previously obtained using `capture`.
+
+If the previous result was a success, this produces its wrapped value.
+If the previous result was an exception, this "rethrows" the exception as if it came
+from where it originated.
+
+`do r ← capture t, unwrap r` is identical to `t`, but allows for intermediate tactics to be inserted.
+-/
+meta def unwrap {α : Type*} (t : tactic_result α) : tactic α :=
+match t with
+| (success r s') := return r
+| e := λ s, e
+end
```

Comment:
```suggestion
λ _, t
```

Then the following both work:
```lean
-- works in 3.21
example {a b : ℕ} : ∃ x, x * a = a * x :=  begin
  existsi _,
  conv {
    rw nat.mul_comm b a,
  },
end

-- fails in 3.21
example {a b : ℕ} : ∃ x, x * a = a * x :=  begin
  existsi _,
  conv {
    for (_ * _) [1] {
      rw nat.mul_comm b a,
    },
  },
end
```

Reply:
That sounds wrong to me - the success tactic_state shouldn't be propagated, otherwise it rolls back any work done between capture and unwrap. Thanks for the examples, I'll have a go with them.

Reply:
Ugh, your suggestion does seem to be the only way to fix those examples, but I have a really hard time believing it's correct behavior - that looks like a much more insidious bug in `ext_simplify_core` that is discarding tactic_state even in successful cases.

Reply:
To be clear - what your suggestion does is completely restore the state to how it was before, which results in all the `tactic_state` changes made between callbacks within `ext_simplify_core` being reverted, even when the callbacks are successful.

Reply:
Yes, that is precisely the point.  Nobody except us is modifying the tactic state, in particular nothing is happening between callbacks (fingers crossed).  And `ext_simplify_core` would just discard the tactic state at the end.  So we need to thread it through manually.

Reply:
> in particular nothing is happening between callbacks (fingers crossed).

I got the feeling from reading the code that the type of thing performed by `simp only` would go on behind the scenes between callbacks. Perhaps we've explicitly disabled every one of these simplifications.

> And ext_simplify_core would just discard the tactic state at the end. 

Is it supposed to do that? My understanding is that it is trying to preserve it, but seems to only be preserving part of it

Reply:
Ah, right - I understand now: 

https://github.com/leanprover-community/lean/blob/42713125a2270a07d7034abfb45af786fb5d8746/src/library/tactic/simplify.cpp#L1284

The state is completely (and deliberately) reset at the end of the loop

Reply:
I don't think too much thought went into the design of `ext_simplify_core`.  To be fair, nobody had much experience with meta code back then.